REGRESSION (292869@main): Selector-based targeting no longer hides elements that are inserted after page load

whsieh opened the pull request:
#### 785a7ceb84b79cb21ea3aad5979a1613bd1b9ede
<pre>
REGRESSION (292869@main): Selector-based targeting no longer hides elements that are inserted after page load
<a href="https://bugs.webkit.org/show_bug.cgi?id=291476">https://bugs.webkit.org/show_bug.cgi?id=291476</a>
<a href="https://rdar.apple.com/148747364">rdar://148747364</a>

Reviewed by Richard Robinson.

In 292869@main, I tried to avoid repeatedly calling `Document::isTopDocument` inside
`adjustVisibilityInRepeatedlyTargetedRegions` by bailing if visibility adjustment had never been
set in the current page. However, this is incorrect, because this method is also responsible for
setting visibility adjustment for elements that match `visibilityAdjustmentSelectors` specified via
website policies.

To fix this, we revert 292869@main and instead replace the check for `isTopDocument` with a check
for whether the given `document`&apos;s frame is equal to the `Page`&apos;s main frame. Importantly, this
avoids hitting `Document::mainFrameDocument()` in this hot codepath, which potentially requires a
frame tree walk.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromSelectorsAfterDelay)):

Add an API test to exercise this change by evaluating JavaScript to show a popup that should be
hidden by website policies, only after a delay following page load.

Canonical link: <a href="https://commits.webkit.org/293645@main">https://commits.webkit.org/293645@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55f728c4c90f5732ea92d16d5848818cdfb567c9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19118 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104599 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50070 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101509 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27551 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75707 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32808 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89815 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56066 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14570 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7799 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49429 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7886 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106957 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26582 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19396 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-webvtt-alignment.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84667 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86019 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84185 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21366 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6555 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20354 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26522 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31723 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26342 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29655 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27909 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
```diff
@@ -430,6 +430,31 @@ @implementation _WKTargetedElementInfo (TestingAdditions)
+        "  popup.style = 'width: 600px; height: 480px; position: fixed; background: tomato; top: 0; left: 0;';"
+        "  document.body.appendChild(popup);"
+        "}, 100);";
+    [webView objectByEvaluatingJavaScript:@(scriptSource)];
```
TIL you can use @() to do const char * -> NSString * !

whsieh committed:
REGRESSION (292869@main): Selector-based targeting no longer hides elements that are inserted after page load
https://bugs.webkit.org/show_bug.cgi?id=291476
rdar://148747364

Reviewed by Richard Robinson.

In 292869@main, I tried to avoid repeatedly calling `Document::isTopDocument` inside
`adjustVisibilityInRepeatedlyTargetedRegions` by bailing if visibility adjustment had never been
set in the current page. However, this is incorrect, because this method is also responsible for
setting visibility adjustment for elements that match `visibilityAdjustmentSelectors` specified via
website policies.

To fix this, we revert 292869@main and instead replace the check for `isTopDocument` with a check
for whether the given `document`'s frame is equal to the `Page`'s main frame. Importantly, this
avoids hitting `Document::mainFrameDocument()` in this hot codepath, which potentially requires a
frame tree walk.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromSelectorsAfterDelay)):

Add an API test to exercise this change by evaluating JavaScript to show a popup that should be
hidden by website policies, only after a delay following page load.

Canonical link: https://commits.webkit.org/293645@main
